src/providers/hana/qgshanaresultset.cpp

```diff
@@ -0,0 +1,142 @@
+      else
+        return QgsHanaUtils::toVariant( mResultSet->getUShort( columnIndex ) );
+    case SQLDataTypes::Integer:
+      if ( mMetadata ->isSigned( columnIndex ) )
+        return QgsHanaUtils::toVariant( mResultSet->getInt( columnIndex ) );
+      else
+        return QgsHanaUtils::toVariant( mResultSet->getUInt( columnIndex ) );
+    case SQLDataTypes::BigInt:
+      if ( mMetadata ->isSigned( columnIndex ) )
+        return QgsHanaUtils::toVariant( mResultSet->getLong( columnIndex ) );
+      else
+        return QgsHanaUtils::toVariant( mResultSet->getULong( columnIndex ) );
+    case SQLDataTypes::Double:
+    case SQLDataTypes::Decimal:
+    case SQLDataTypes::Numeric:
+      return QgsHanaUtils::toVariant( mResultSet->getDouble( columnIndex ) );
+    case SQLDataTypes::Date:
+    case SQLDataTypes::TypeDate:
+      return QgsHanaUtils::toVariant( mResultSet->getDate( columnIndex ) );
+    case SQLDataTypes::Time:
+    case SQLDataTypes::TypeTime:
+      return QgsHanaUtils::toVariant( mResultSet->getTime( columnIndex ) );
+    case SQLDataTypes::Timestamp:
+    case SQLDataTypes::TypeTimestamp:
+      return QgsHanaUtils::toVariant( mResultSet->getTimestamp( columnIndex ) );
+    case SQLDataTypes::VarChar:
+    case SQLDataTypes::LongVarChar:
+      return QgsHanaUtils::toVariant( mResultSet->getString( columnIndex ) );
+    case SQLDataTypes::WVarChar:
+    case SQLDataTypes::WLongVarChar:
+      return QgsHanaUtils::toVariant( mResultSet->getNString( columnIndex ) );
+    case SQLDataTypes::Binary:
+    case SQLDataTypes::VarBinary:
+    case SQLDataTypes::LongVarBinary:
+      return QgsHanaUtils::toVariant( mResultSet->getBinary( columnIndex ) );
+    default:
+      QgsDebugMsg( QStringLiteral( "Unhandled HANA type %1" ).arg( QString::fromStdString( mMetadata->getColumnTypeName( columnIndex ) ) ) );
+      return QVariant();
+  }
+}
+
+QgsGeometry QgsHanaResultSet::getGeometry( unsigned short columnIndex )
+{
+  size_t bufLength = mResultSet->getBinaryLength( columnIndex );
+  unsigned char *bufPtr = nullptr;
+
+  if ( bufLength == ResultSet::UNKNOWN_LENGTH )
+  {
+    Binary wkb = mResultSet->getBinary( columnIndex );
+    if ( wkb.isNull() || wkb->size() == 0 )
+      bufLength = 0;
+    else
+      bufPtr = reinterpret_cast<unsigned char *>( wkb->data() );
+  }
+  else if ( bufLength != 0 && bufLength != odbc::ResultSet::NULL_DATA )
+  {
+    ensureBufferCapacity( bufLength );
+    mResultSet->getBinaryData( columnIndex, mBuffer.data(), bufLength );
+    bufPtr = mBuffer.data();
+  }
+
+  QgsGeometry geom;
+  if ( !( bufLength == 0 || bufPtr == nullptr ) )
+  {
+    unsigned char *wkbGeom = new unsigned char[bufLength];
+    memcpy( wkbGeom, bufPtr, bufLength );
+    geom.fromWkb( wkbGeom, static_cast<int>( bufLength ) );
+  }
+
+  return geom;
+}
+
+void QgsHanaResultSet::ensureBufferCapacity( std::size_t capacity )
+{
+  if ( capacity > mBuffer.size() )
+  {
+    mBuffer.reserve( capacity );
+  }
```

Comment:
`reserve()` is not sufficient if you write to the vector via `data()`. You have to call `resize()` here. Actually, you don't need that method at all. You can just call `resize()` before writing to the buffer. Already allocated memory won't be freed. So there are no additional allocations if you make the vector smaller and restore the former size later.

Reply:
Done in 4d36f6f.